airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/SSHTunnel.java

```diff
@@ -0,0 +1,272 @@
+  public void openTunnelIfRequested() throws IOException {
+    if (shouldTunnel()) {
+      if (tunnelSession != null || sshclient != null) {
+        throw new RuntimeException("SSH Tunnel was requested to be opened while it was already open.  This is a coding error.");
+      }
+      sshclient = createClient();
+      tunnelSession = openTunnel(sshclient);
+    }
+  }
+
+  /**
+   * Closes a tunnel if one was open, and otherwise doesn't do anything (safe to run).
+   */
+  public void closeTunnel() {
+    try {
+      if (shouldTunnel()) {
+        if (tunnelSession != null) {
+          tunnelSession.close();
+        }
+        if (sshclient != null) {
+          sshclient.stop();
+        }
+        tunnelSession = null;
+        sshclient = null;
+      }
+    } catch (Throwable t) {
+      throw new RuntimeException(t);
+    }
+  }
+
+  public boolean shouldTunnel() {
+    return method != null && !"NO_TUNNEL".equals(method);
+  }
+
+  public String getMethod() {
+    return method;
+  }
+
+  public String getHost() {
+    return host;
+  }
+
+  public String getTunnelSshPort() {
+    return tunnelSshPort;
+  }
+
+  public String getUser() {
+    return user;
+  }
+
+  private String getSSHKey() {
+    return sshkey;
+  }
+
+  private String getPassword() {
+    return password;
+  }
+
+  public String getRemoteDatabaseHost() {
+    return remoteDatabaseHost;
+  }
+
+  public String getRemoteDatabasePort() {
+    return remoteDatabasePort;
+  }
+
+  public String getTunnelDatabasePort() {
+    return tunnelDatabasePort;
+  }
+
+  /**
+   * From the RSA format private key string, use bouncycastle to deserialize the key pair, reconstruct
+   * the keys from the key info, and return the key pair for use in authentication.
+   *
+   * @return
+   * @throws IOException
+   */
+  private KeyPair getPrivateKeyPair() throws IOException {
+    PEMParser pemParser = new PEMParser(new StringReader(getSSHKey()));
+    PEMKeyPair keypair = (PEMKeyPair) pemParser.readObject();
+    JcaPEMKeyConverter converter = new JcaPEMKeyConverter();
+    return new KeyPair(
+        (RSAPublicKey) converter.getPublicKey(SubjectPublicKeyInfo.getInstance(keypair.getPublicKeyInfo())),
+        (RSAPrivateKey) converter.getPrivateKey(keypair.getPrivateKeyInfo()));
+  }
+
+  /**
+   * Generates a new ssh client and returns it, with forwarding set to accept all types; use this
+   * before opening a tunnel.
+   *
+   * @return
+   */
+  private SshClient createClient() {
+    java.security.Security.addProvider(
+        new org.bouncycastle.jce.provider.BouncyCastleProvider());
+    SshClient client = SshClient.setUpDefaultClient();
+    client.setForwardingFilter(AcceptAllForwardingFilter.INSTANCE);
+    client.setServerKeyVerifier(AcceptAllServerKeyVerifier.INSTANCE);
+    return client;
+  }
+
+  private void validate() {
+    if (getHost() == null) {
+      throw new RuntimeException("SSH Tunnel host is null - verify configuration before starting tunnel!");
+    }
```

Comment:
```suggestion
    Preconditions.notNull(getHost(), "SSH Tunnel host is null - verify configuration before starting tunnel!")
```
we tend to use Preconditions for validations like this. 1. theoretically takes fewer lines of code (though that's not always true) 2. gives you a more specific error message (in this case NPE)